apps/vscode/src/providers/editor/editor.ts

```diff
@@ -139,8 +143,52 @@ export class VisualEditorProvider implements CustomTextEditorProvider {
       }
     }));
 
+    context.subscriptions.push(window.tabGroups.onDidChangeTabs(async (t) => {
+      const tabs = t.opened;
+
+      if (tabs.length > 0) {
+        for (const tab of tabs) {
+          if (tab.label.endsWith(".qmd") && (tab.input instanceof TabInputText || tab.input instanceof TabInputCustom)) {
+            // determine what mode editor should be in
+            const uri = tab.input.uri;
+
+            const isTextEditor = tab.input instanceof TabInputText;
+            const viewType = isTextEditor ? "textEditor" : tab.input.viewType;
+
+            // get file contents
+            const fileData = await workspace.fs.readFile(uri);
+            const fileContent = Buffer.from(fileData).toString('utf8');
+            const editorMode = determineMode(fileContent, uri);
+            let isSwitch = this.visualEditorPendingSwitchToSource.has(uri.toString()) || this.editorPendingSwitchToVisual.has(uri.toString());
+            if (this.editorPendingSwitchToVisual.has(uri.toString())) {
+              this.editorPendingSwitchToVisual.delete(uri.toString());
+            }
+
+            if (editorMode && editorMode != viewType && !isSwitch) {
+              const allTabs = window.tabGroups.all?.[0]?.tabs;
```

Comment:
I think we need a big meaningful comment here, because this is (at first glance) _really_ weird that we have to do this.

In my experience playing around with this if you set `await window.tabGroups.close(tab, true);` using the `tab` you get from the change event, it won't close and you get a "tab does not exist" error in the debug console.

From what I can tell, the `tab` we get in the change event is not PRECISELY equal to the tabs that exist in `window.tabGroups` at this point in time. It's possible that during this `opened` event they are a little out of sync. The tab we care about closing does exist in there, we just have to find it by matching by URI - something we know is stable across `change` events.

So I think it looks like this:

```
// Unfortunately we cannot just `await window.tabGroups.close(tab, true);`.
// The `tab` we get from the change event seems to not be precisely the same
// as the tab in `window.tabGroups`, so if we try and close `tab` we get a "tab
// not found" error. The one we care about does exist in `window.tabGroups`,
// we just have to manually find it using the URI as a stable field to match on.
```